pkg/controller/provider/web/openstack/workload.go

```diff
@@ -95,36 +95,175 @@ func (r *Workload) Link(p *api.Provider) {
+	if r.ImageID != "" {
+		imageID = r.ImageID
+	}
+	flavor := model.Flavor{Base: model.Base{
+		ID: r.FlavorID,
 	}}
-	db.Get(&image)
-	r.Image.With(&image)
+	err = db.Get(&flavor)
+	if err != nil {
+		return
+	}
+	r.Flavor.With(&flavor)
+
+	var networks []Network
+	var subnets []Subnet
+
+	for name := range r.Addresses {
+		networkList := []model.Network{}
+		err = db.List(&networkList, model.ListOptions{
+			Predicate: libmodel.Eq("Name", name),
+			Detail:    model.MaxDetail,
+		})
+		if err != nil {
+			return
+		}
+		for _, networkModel := range networkList {
+			network := &Network{}
+			network.With(&networkModel)
+			networks = append(networks, *network)
+			for _, subnetID := range network.Subnets {
+				subnetModel := model.Subnet{Base: model.Base{
+					ID: subnetID,
+				}}
+				err = db.Get(&subnetModel)
+				if err != nil {
+					return
+				}
+				subnet := &Subnet{}
+				subnet.With(&subnetModel)
+				subnets = append(subnets, *subnet)
+				if err != nil {
+					return
+				}
+			}
+		}
+	}
+	r.Networks = networks
+	r.Subnets = subnets
 
 	var volumes []Volume
-	for _, av := range r.VM.AttachedVolumes {
+	var volumeTypes []VolumeType
+	var snapshots []Snapshot
+
+	volumeTypeCache := map[string]VolumeType{}
+
+	for _, attachedVolume := range r.AttachedVolumes {
+		volume := &Volume{}
 		volumeModel := model.Volume{
-			Base: model.Base{ID: av.ID},
+			Base: model.Base{ID: attachedVolume.ID},
+		}
+		err = db.Get(&volumeModel)
+		if err != nil {
+			return
 		}
-		db.Get(&volumeModel)
-		volume := &Volume{}
 		volume.With(&volumeModel)
 		volumes = append(volumes, *volume)
+
+		if _, ok := volumeTypeCache[volume.VolumeType]; !ok {
+			volumeTypes := []model.VolumeType{}
+			err = db.List(&volumeTypes, model.ListOptions{
+				Predicate: libmodel.Eq("Name", volume.VolumeType),
+				Detail:    model.MaxDetail,
+			})
+			if err != nil {
+				return
+			}
+			volumeType := &VolumeType{}
+			volumeType.With(&volumeTypes[0])
+			volumeTypeCache[volume.VolumeType] = *volumeType
+		}
+
+		volumeType := volumeTypeCache[volume.VolumeType]
+		found := false
+		for i := range volumeTypes {
+			if volumeTypes[i].ID == volumeType.ID {
+				found = true
+			}
+		}
+		if !found {
+			volumeTypes = append(volumeTypes, volumeType)
+		}
+
+		snapshotsList := []model.Snapshot{}
+		err = db.List(&snapshotsList, model.ListOptions{
+			Predicate: libmodel.Eq("volumeID", attachedVolume.ID),
+			Detail:    model.MaxDetail,
+		})
+		if err != nil {
+			return
+		}
+		for _, snapshotModel := range snapshotsList {
+			snapshot := &Snapshot{}
+			snapshot.With(&snapshotModel)
+			snapshots = append(snapshots, *snapshot)
+		}
+
+		if imageID == "" && volume.Bootable == "true" {
+			if volumeImageID, ok := volume.VolumeImageMetadata["image_id"]; ok {
+				imageID = volumeImageID
+			}
+		}
+
+	}
+
+	r.Volumes = volumes
+	r.VolumeTypes = volumeTypes
+	r.Snapshots = snapshots
+
+	if imageID != "" {
+		image := model.Image{Base: model.Base{
+			ID: imageID,
```

Comment:
```suggestion
			ID: imageID,
			TypeID: volumeType,
```
this will be nice for the UI

Reply:
The volume type is related to volumes, not images. If you are retrieving the workload information use `volumes` instead of `attachedVolumes` and you will get all the information about the volume, not only the `ID`.

Reply:
ok, in current version we only have attachedVolumes

https://github.com/kubev2v/forklift-console-plugin/pull/343/files#diff-9e59aeb6e8ebb0803d37926497b9ef77244ffdd8e597c56b2f2695830a7e17feR46

if we also get the full volumes data when getting the vms that will be great

Reply:
to clarify what the UI need this data for:
when showing the user the posible storage maps,
the UI look at the vm, and show the storage maps that have matching storage type for the volume types of the vm attached volumes.

ovirt and vmware give the storage type for attached volumes in the vm api call, for that use case.

for openstack to get this behavior, we need:
get the vm,
get all volumes, find the volume with matching id.
get the volume type of that volume, and add it to the vm information

then we can match the storage mappings that work with this vm.